statistics/handle/lru_cache.go

```diff
@@ -0,0 +1,334 @@
+	if capacity < 1 {
+		panic("capacity of LRU Cache should be at least 1.")
+	}
+	return &internalLRUCache{
+		capacity: capacity,
+		elements: make(map[int64]*list.Element),
+		cache:    list.New(),
+	}
+}
+
+// Get tries to find the corresponding value according to the given key.
+func (l *internalLRUCache) Get(key int64) (*statistics.Table, bool) {
+	var hit bool
+	var r *statistics.Table
+	l.Lock()
+	r, hit = l.get(key)
+	l.Unlock()
+	if hit {
+		metrics.StatsCacheLRUCounter.WithLabelValues(typHit).Inc()
+	} else {
+		metrics.StatsCacheLRUCounter.WithLabelValues(typMiss).Inc()
+	}
+	return r, hit
+}
+
+func (l *internalLRUCache) get(key int64) (*statistics.Table, bool) {
+	element, exists := l.elements[key]
+	if !exists {
+		return nil, false
+	}
+	l.cache.MoveToFront(element)
+	return element.Value.(*cacheItem).value, true
+}
+
+// Put puts the (key, value) pair into the LRU Cache.
+func (l *internalLRUCache) Put(key int64, value *statistics.Table) bool {
+	var success bool
+	var trackingCost int64
+	var totalCost int64
+	l.Lock()
+	success = l.put(key, value, value.MemoryUsage(), true)
+	trackingCost = l.trackingCost
+	totalCost = l.totalCost
+	l.Unlock()
+	if success {
+		metrics.StatsCacheLRUCounter.WithLabelValues(typUpdate).Inc()
+		metrics.StatsCacheLRUMemUsage.WithLabelValues(typTrack).Set(float64(trackingCost))
+		metrics.StatsCacheLRUMemUsage.WithLabelValues(typTotal).Set(float64(totalCost))
+	}
+	return success
+}
+
+func (l *internalLRUCache) put(key int64, value *statistics.Table, tblMemUsage *statistics.TableMemoryUsage, tryEvict bool) bool {
+	if l.capacity < tblMemUsage.TotalColTrackingMemUsage() {
+		return false
+	}
+	defer func() {
+		if tryEvict {
+			l.evictIfNeeded()
+		}
+	}()
+	element, exists := l.elements[key]
+	if exists {
+		oldMemUsage := element.Value.(*cacheItem).tblMemUsage
+		element.Value.(*cacheItem).value = value
+		element.Value.(*cacheItem).tblMemUsage = tblMemUsage
+		l.calculateCost(tblMemUsage, oldMemUsage)
+		l.maintainList(element, nil, tblMemUsage, oldMemUsage)
+		return true
+	}
+	newCacheEntry := &cacheItem{
+		key:         key,
+		value:       value,
+		tblMemUsage: tblMemUsage,
+	}
+	l.calculateCost(tblMemUsage, &statistics.TableMemoryUsage{})
+	element = l.maintainList(nil, newCacheEntry, tblMemUsage, &statistics.TableMemoryUsage{})
+	l.elements[key] = element
+	return true
+}
+
+// Del deletes the key-value pair from the LRU Cache.
+func (l *internalLRUCache) Del(key int64) {
+	var del bool
+	var trackingCost int64
+	var totalCost int64
+	l.Lock()
+	del = l.del(key)
+	trackingCost = l.trackingCost
+	totalCost = l.totalCost
+	l.Unlock()
+	if del {
+		metrics.StatsCacheLRUCounter.WithLabelValues(typDel).Inc()
+		metrics.StatsCacheLRUMemUsage.WithLabelValues(typTrack).Set(float64(trackingCost))
+		metrics.StatsCacheLRUMemUsage.WithLabelValues(typTotal).Set(float64(totalCost))
+	}
+}
+
+func (l *internalLRUCache) del(key int64) bool {
+	element := l.elements[key]
```

Comment:
Maybe `element, exists := l.elements[key]` is better?

Reply:
updated